WordPress/Classes/Utility/BackgroundTasks/WeeklyRoundupBackgroundTask.swift

```diff
@@ -34,89 +35,154 @@ private class WeeklyRoundupDataProvider {
         self.onError = onError
     }
 
+    // MARK: API
+
+    /// Fetches the top site statistics from all available sites in the provided context.
+    ///
+    /// This method retrieves all sites in the given context, then fetches the weekly stats
+    /// for each site. The result, which includes the top 5 sites based on the received stats,
+    /// is returned through the provided completion handler.
+    ///
+    /// **Note:** The completion handler is executed on a background queue.
     func getTopSiteStats(completion: @escaping (Result<SiteStats?, Error>) -> Void) {
-        getSites() { [weak self] sitesResult in
+        self.coreDataStack.performAndSave { [weak self] context in
+            guard let self else {
+                completion(.success(nil))
+                return
+            }
+            self.getTopSiteStats(in: context, completion: completion)
```

Comment:
This context intance is passed on to many functions, most of which runs asynchronously. There are two issues I can see here:

First, it's best that we don't let the context object created by `performAndSave` to live outside of the `performAndSave` scope: When `performAndSave` exits, we should stop using the context object. Otherwise, we can't ensure changes made to the context object is saved by `performAndSave`.

Second, I don't think we _need_ to use one single context object across that many functions. For example, `fetchStats(for:endingOn:in:with:completion:)` don't need a `context` argument. It can accepts an site id argument instead of a `Blog` instance. `filterWeeklyRoundupEnabledSites(_:in:result:)` also don't need a `context` argument. It can accepts a list of `NSManagedObjectID` and returns a list of `NSManagedObjectID` to the result closure. The same changes can probably be made to most of the functions that takes a `context` argument.

Reply:
> First, it's best that we don't let the context object created by performAndSave to live outside of the performAndSave scope: When performAndSave exits, we should stop using the context object. Otherwise, we can't ensure changes made to the context object is saved by performAndSave.

Good point 👍

Reply:
Now, as for what I ended up implementing, when fetching the list of `Blogs`, I mapped the result to a plain, context-free `Site` struct. Like this:

```swift
private struct Site {
    let managedObjectID: NSManagedObjectID
    let authToken: String?
    let dotComID: Int?
    let timeZone: TimeZone
    let isAdmin: Bool
    let isAutomatticP2: Bool

    init(blog: Blog) {
        self.managedObjectID = blog.objectID
        self.authToken = blog.account?.authToken
        self.dotComID = blog.dotComID?.intValue
        self.timeZone = blog.timeZone
        self.isAdmin = blog.isAdmin
        self.isAutomatticP2 = blog.isAutomatticP2
    }
}

private func getAllSites(in context: NSManagedObjectContext) -> Result<[Site], Error> {
    // ...
    let result = try context.fetch(request)
    let sites = result.map { Site(blog: $0) }
    // ...
}
```

Reply:
I realize this solution isn't very elegant, but if there aren't any major issues, I propose we proceed with this Pull Request. We can refine and improve it in future Pull Requests. 🙏

Reply:
That looks awesome. The functions look super clean. Thanks for making the changes!